src/clr-angular/data/tree-view/recursive-for-of.ts

```diff
@@ -1,25 +1,27 @@
 /*
- * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
  * This software is released under MIT license.
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 
-import { Directive, Input, OnChanges, TemplateRef } from '@angular/core';
+import { ChangeDetectorRef, Directive, Input, OnChanges, OnDestroy, TemplateRef } from '@angular/core';
 import { TreeFeaturesService } from './tree-features.service';
 import { RecursiveTreeNodeModel } from './models/recursive-tree-node.model';
 import { TreeNodeModel } from './models/tree-node.model';
 import { AsyncArray } from './models/async-array';
+import { Subscription } from 'rxjs';
 
 export interface ClrRecursiveForOfContext<T> {
   $implicit: T;
   clrModel: TreeNodeModel<T>;
 }
 
 @Directive({ selector: '[clrRecursiveFor][clrRecursiveForOf]' })
-export class ClrRecursiveForOf<T> implements OnChanges {
+export class ClrRecursiveForOf<T> implements OnChanges, OnDestroy {
   constructor(
     private template: TemplateRef<ClrRecursiveForOfContext<T>>,
-    private featuresService: TreeFeaturesService<T>
+    private featuresService: TreeFeaturesService<T>,
+    private cdr: ChangeDetectorRef
```

Comment:
This seems like a `Chocolate error`(™). Did you try to fix this error following the willy wonka/oompa-loompa pattern?

Reply:
At first it looked like a very good idea, but after I looked in detail how to do it I think it's not applicable in this case. Why?
1) The `recursive-for-of` directive may be a good candidate for the Willy Wonka half of the task, but
2) The Oompa-loompa part has to be implemented in the `recursive-tree-node-model` which, being a model, is a bit outside of the ordinary angular lifecycle. That's why I need to trigger change detection indirectly through a service anyway.
We can discuss it in more detail offline.

Reply:
Okay, we had an offline discussion and did some tests with Matt (thanks!) and here is the summary:
- willy wonka/oompa-loompa pattern is unsuitable because one of the classes involved is a model class, that knows nothing about Angular.
- Tried to substitute the Subject instance in the featuresService with first an `Injector` and second with a `ChangeDetectorRef` but in both cases ended with test nightmares, and the model is used as a model in the tests (and should be) and there are many tests where generated instances are compared with manually created sample instances. If these models include Angular classes, it's very hard to supply the proper dependencies. This is not only hard to test, but tells us that we've abused the class meant to be a model class and in general that our approach is not really correct.
So, the initial implementation seems to be the best choice of all evaluated, only we have to make sure that we'll not leak subscriptions in the case that ngOnChanges is triggered more than once (which is unlikely but not impossible).